sql/src/main/java/io/crate/analyze/relations/RelationAnalyzer.java

```diff
@@ -90,18 +92,28 @@ protected AnalyzedRelation visitUnion(Union node, StatementAnalysisContext conte
     }
 
     @Override
-    protected AnalyzedRelation visitJoin(Join node, StatementAnalysisContext context) {
+    protected AnalyzedRelation visitJoin(Join node, StatementAnalysisContext statementContext) {
         if (!ALLOWED_JOIN_TYPES.contains(node.getType())) {
             throw new UnsupportedOperationException("Explicit " + node.getType().name() + " join syntax is not supported");
         }
-        process(node.getLeft(), context);
-        process(node.getRight(), context);
+        process(node.getLeft(), statementContext);
+        process(node.getRight(), statementContext);
+
+        RelationAnalysisContext relationContext = statementContext.currentRelationContext();
 
         Optional<JoinCriteria> optCriteria = node.getCriteria();
         if (optCriteria.isPresent()) {
             JoinCriteria joinCriteria = optCriteria.get();
             if (joinCriteria instanceof JoinOn) {
-                context.currentRelationContext().addJoinExpression(((JoinOn) joinCriteria).getExpression());
+                Symbol joinCondition;
+                try {
+                    joinCondition = relationContext.expressionAnalyzer().convert(
+                        ((JoinOn) joinCriteria).getExpression(), relationContext.expressionAnalysisContext());
+                } catch (RelationUnknownException e) {
+                    throw new ValidationException(String.format(Locale.ENGLISH,
+                        "missing FROM-clause entry for relation '%s'", e.qualifiedName()));
+                }
+                statementContext.currentRelationContext().addJoinCondition(joinCondition);
```

Comment:
is currentRelationContext() not the same as relationContext here? Could use the local variable